specs/ProgrammaticSaveAs.md

```diff
@@ -0,0 +1,267 @@
+Programmatic Save As API
+===
+
+# Background
+
+The context menu has the "Save as" item to manually save the html page, image, pdf, or other content through a save as dialog. We provide more flexiable ways to do the save as programmatically in WebView2. You can bring up the default save as dialog easily. And you will be able to block default dialog, save the content silently, by providing the path and save as type programmatically or even build your own save as UI.
+
+In this document we describe the API. We'd appreciate your feedback.
+
+# Description
+
+We propose the `SaveContentAs` method WebView2, which allows you to trigger the save as programmatically. By using this method alone, the system default will popup.
+
+Additionally, we propose the `SaveAsRequestedEvent`. You can register this event to block the default dialog and use the `SaveAsRequestedEventArgs` instead, to set your perferred save as path, save as type, and depulicate file replacement rule. In your clinet app, you can design your own UI to input these parameters. For html page, we support 3 save as types: HTML_ONLY, SINGLE_FILE and COMPLETE. For non-html page, the type is been set as DEFAULT, which will save the content as it is. This API also provides default values for all parameters, if you don't want to input anything.
+
+# Examples
+## Win32 C++
+```c++
+//! [ToggleSilent]
+// Turn on/off Silent SaveAs, which won't show the system default save as dialog.
+// This example hides the default save as dialog and shows a customized dialog.
+bool ScenarioSaveAs::ToggleSilent()
+{
+    if (!m_webView2Staging20)
+        return false;
+    if (m_silentSaveAs)
+    {
+        // Unregister the handler for the `SaveAsRequested` event.
+        m_webView2Staging20->remove_SaveAsRequested(m_saveAsRequestedToken);
+    }
+    else
+    {
+        // Register a handler for the `SaveAsRequested` event.
+        m_webView2Staging20->add_SaveAsRequested(
+            Callback<ICoreWebView2StagingSaveAsRequestedEventHandler>(
+                [this](
+                    ICoreWebView2* sender,
+                    ICoreWebView2StagingSaveAsRequestedEventArgs* args) -> HRESULT
+                {
+                    // Hide the system default save as dialog.
+                    CHECK_FAILURE(args->put_Handled(TRUE));
+
+                    auto showCustomizedDialog = [this, args]
+                    {
+                        // Preview the content mime type, optional
+                        wil::unique_cotaskmem_string mimeType;
+                        CHECK_FAILURE(args->get_ContentMimeType(&mimeType));
+
+                        SaveAsDialog dialog(m_appWindow->GetMainWindow(), contentSaveTypes);
```

Comment:
A real app replacing the save as dialog would probably want to use the name of the file and file type and maybe other info in their dialog. If we're not going to do this in the sample, please at least include a comment explaining that the end developer may want to pass the event arg info here so that it can be used in the replacement dialog.